pkg/apis/crd/v1alpha1/types.go

```diff
@@ -291,3 +291,157 @@ type TLSProtocol struct {
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+}
+
+// TCPHeader describes spec of a TCP header.
+type TCPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
+	// Flags are flags in the header.
+	Flags int32 `json:"flags,omitempty"`
+}
+
+// Packet includes header info.
+type Packet struct {
+	SrcIP string `json:"srcIP,omitempty"`
+	DstIP string `json:"dstIP,omitempty"`
+	// Length is the IP packet length (includes the IPv4 or IPv6 header length).
+	Length          int32           `json:"length,omitempty"`
+	IPHeader        IPHeader        `json:"ipHeader,omitempty"`
+	IPv6Header      *IPv6Header     `json:"ipv6Header,omitempty"`
+	TransportHeader TransportHeader `json:"transportHeader,omitempty"`
+}
+
+type PacketCaptureType string
+
+const (
+	FirstNCapture PacketCaptureType = "FirstNCapture"
+)
+
+type FirstNCaptureConfig struct {
+	Number int32 `json:"number,omitempty"`
+}
+
+const DefaultPacketCaptureTimeout uint16 = 60
+
+type PacketCapturePhase string
+
+const (
+	PacketCaptureRunning   PacketCapturePhase = "Running"
+	PacketCaptureSucceeded PacketCapturePhase = "Succeeded"
+	PacketCaptureFailed    PacketCapturePhase = "Failed"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCaptureList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	Items []PacketCapture `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+type PacketCapture struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              PacketCaptureSpec   `json:"spec,omitempty"`
+	Status            PacketCaptureStatus `json:"status,omitempty"`
+}
+
+type PacketCaptureSpec struct {
+	Timeout uint16 `json:"timeout,omitempty"`
+	// Type is the capture type. Currently only FirstN is supported.
+	Type PacketCaptureType `json:"type,omitempty"`
+	// FirstNCaptureConfig contains the config for the FirstN type capture. The only supported parameter is
+	// `Number` at the moment, meaning capturing the first specified number of packets in a flow.
+	FirstNCaptureConfig *FirstNCaptureConfig `json:"firstNCaptureConfig,omitempty"`
+	Source              Source               `json:"source,omitempty"`
+	Destination         Destination          `json:"destination,omitempty"`
+	Packet              Packet               `json:"packet,omitempty"`
+	// FileServer specifies the sftp url config for the fileServer. Captured packets will be uploaded to this server.
+	FileServer     BundleFileServer              `json:"fileServer,omitempty"`
+	Authentication BundleServerAuthConfiguration `json:"authentication,omitempty"`
```

Comment:
My understanding is that using `omitempty` with struct fields is misleading. If the field is optional, it should be a pointer, If it is not optional, we should not have the `omitempty` tag. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required. cc @tnqn for his opinion.

Reply:
Agree, `omitempty` should be removed